### DISCUSSION TITLE
Deterministic order for bulk mutation that uses multiple different remote schemas?

### DISCUSSION THREAD

**Discussion by kevinmichaelchen** [hasura/graphql-engine#9474] [General] [answered]
When performing a bulk mutation to multiple different remote schemas, is there a deterministic order of execution?

**Comment by BenoitRanque** [comment, marked as answer]
From the [GraphQL spec](https://spec.graphql.org/October2021/#sec-Mutation), emphasis mine:

> If the operation is a mutation, the result of the operation is the result of executing the operation’s top level selection set on the mutation root object type. **_This selection set should be executed serially_**.
>
> It is expected that the top level fields in a mutation operation perform side-effects on the underlying data system. Serial execution of the provided mutations ensures against race conditions during these side-effects.

So yes, the root fields of a mutation are executed serially, and thus have a deterministic order of execution.

This of course assumes the remote schemas won't return a result until they are finished executing.

If the remote schemas return a result before they are finished executing whatever work the mutation entails, then you may face a race condition.
